Utils/src/PDFWeightProducer.cc

```diff
@@ -63,57 +123,84 @@ void PDFWeightProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::Ev
-    
-    std::vector< gen::WeightsInfo > lheweights = LheInfo->weights();
-    if(!lheweights.empty()){
-      found_weights = true;
-      // these numbers are hard-coded by the LHEEventInfo
-      //renormalization/factorization scale weights
-      for (unsigned int i = 0; i < 9; i++){
-        scaleweights.push_back(lheweights[i].wgt/lheweights[0].wgt);
-      }
-      //pdf weights
-      for (unsigned int i = 9; i < 110; i++){
-        pdfweights.push_back(lheweights[i].wgt/lheweights[9].wgt);
-        pdfids.push_back(i-9);
-      }
-    }
+    auto helper = makeHelper(LheInfo.product());
+    //renormalization/factorization scale weights
+    found_scales = helper.fillWeights(scaleweights.get(),0,nScales_,false);
+    //pdf weights
+    if(found_scales) found_pdfs = helper.fillWeights(pdfweights.get(),nScales_,nPDFs_,true);
   }
   
   //check GenEventInfoProduct if LHEEventProduct not found or empty
-  if(!found_weights){
-    edm::Handle<GenEventInfoProduct> genHandle;
-    iEvent.getByToken(genProductToken_, genHandle);
-
-	const std::vector<double>& genweights = genHandle->weights();
-    // these numbers are hard-coded by the GenEventInfo (shifted by 1 wrt LHE)
-    //renormalization/factorization scale weights
-    for (unsigned int i = 1; i < 10; i++){
-      scaleweights.push_back(genweights[i]/genweights[1]);
+  //if LHEEventProduct was filled, check GenEventInfoProduct for parton shower weights from Pythia8
+  edm::Handle<GenEventInfoProduct> genHandle;
+  iEvent.getByToken(genProductToken_, genHandle);
+  if(genHandle.isValid()){
+    auto helper = makeHelper(genHandle.product());
+    if(!found_scales or !found_pdfs){
+      unsigned offset = 1;
+      //renormalization/factorization scale weights
+      found_scales = helper.fillWeights(scaleweights.get(),offset,nScales_,false);
+      //pdf weights
+      if(found_scales) found_pdfs = helper.fillWeights(pdfweights.get(),nScales_+offset,nPDFs_,true);
     }
-    //pdf weights
-    for (unsigned int i = 10; i < 111; i++){
-      pdfweights.push_back(genweights[i]/genweights[10]);
-      pdfids.push_back(i-10);
+    else if(!found_pss){
+      unsigned offset = 2;
+      found_pss = helper.fillWeights(psweights.get(),offset,nPSs_,false);
+    }
+
+    //if pdf weights still not found, recalculate them
+    //taken from https://github.com/cms-sw/cmssw/blob/master/ElectroWeakAnalysis/Utilities/src/PdfWeightProducer.cc
+    //not sure what to do about missing scale weights
+    if(!found_pdfs and recalculatePDFs_){
+      float Q = genHandle->pdf()->scalePDF;
+
+      int id1 = genHandle->pdf()->id.first;
+      double x1 = genHandle->pdf()->x.first;
+      double pdf1 = genHandle->pdf()->xPDF.first;
+
+      int id2 = genHandle->pdf()->id.second;
+      double x2 = genHandle->pdf()->x.second;
+      double pdf2 = genHandle->pdf()->xPDF.second;
+      if(pdf1 <= 0 && pdf2 <= 0) {
+         std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
What is the mutex_ doing? I understand that you are probably making the code thread safe, but I see a lock and no corresponding unlock. FYI, I have never use a mutex and have no idea what I'm talking about. Just trying to understand the code a little better.

Reply:
`mutex_` is a static member variable of the class. If it is locked in one thread, other threads will wait to execute this part of the code until it is unlocked. The lock is created within the scope of the if statement, so once it goes out of scope and destructs, the mutex is unlocked.

Reply:
Got it. Thanks for the explanation.